### ISSUE TITLE
MacOS: Starting RTSP handshake failed: Error 50

### ISSUE THREAD

**Issue by mccoy-pauley-df** [moonlight-stream/moonlight-qt#587]
Everything was working beautifully: Windows 10 host with GeForce RTX 3070, GeForce Experience 3.21.0.36, MacOs Big Sur client with Moonlight-qt 3.1.1, both PCs on local 1Gb Ethernet LAN on same subnet.  After GFE automatically upgraded to 3.22.0.32, and upgrading MacOS client to 3.1.2, I consistently get "Starting RTSP Handshake failed: Error 50" when trying to start or connect to any session (mostly trying Playnite or mstsc.exe).  Tried reinstalling, then downgrading GFE and Moonlight-qt, but problem persists across reinstalls and reboots. Confirmed all firewalls are open and observed traffic with Wireshark, which shows no sign of blocked connections. Tried removing Moonlight prefs plist ~/Library/Preferences/com.moonlight-stream.Moonlight.plist and restarting/reconfiguring from scratch, but no change.

Moonlight log seems to indicate successful conversations, but failure with RTSP options request.

00:00:11 - SDL Info (0): Starting RTSP handshake...
00:00:11 - SDL Info (0): Failed to read RTSP response: 50
00:00:11 - SDL Info (0): RTSP OPTIONS request failed: 50
00:00:11 - SDL Info (0): failed: 50

Moonlight iOS client 7.0.7 on iPadOS 14.5.1 connects to the same host without any issues.

Windows Host: 
Win10 19042.064 (version 20H2)
NVidia drivers 466.27
Antivirus: Sophos Home

MacOS client:
MacBook Pro 15" 2018 model
Radeon Pro 560X 4GB
macOS Big Sur version 11.3.1

Have not seen any other reports of error 50.  Moonlight log and error screenshot are attached.

<img width="522" alt="Screen Shot 2021-05-05 at 2 28 56 PM" src="https://user-images.githubusercontent.com/83726640/117194472-21ee9580-adb2-11eb-9ee5-d77c9900ed3b.png">
[Moonlight-1620239990.log](https://github.com/moonlight-stream/moonlight-qt/files/6429590/Moonlight-1620239990.log)

Am uncertain if some outdated info is being cached by the client which causes miscommunication, but not sure what to clear beyond the prefs plist.  Thanks for any help!

**Comment by cgutman** [comment]
That's quite interesting. Error 50 is ENETDOWN:
```
/* ipc/network software -- operational errors */
#define ENETDOWN        50              /* Network is down */
```

That's a rather odd error code to get back from a call to `recv()` on a connected socket. Clearly the network can't be down if we've got a live socket connected. Is there any chance you've got some firewall (macOS Firewall, Little Snitch, etc.) enabled on your client?

Do you mind uploading a Wireshark packet capture? You can use the filter `tcp.port == 48010` to isolate traffic to just the RTSP connection.

**Comment by mccoy-pauley-df** [comment]
Sure thing! gzipped PCAP is attached, filtered to TCP 48010.  .33 is the client, .35 is the server.  Seems like a fairly straightforward conversation.  The native MacOS firewall is enabled, but no changes have been made there since it was working, and Moonlight is specifically enabled to allow incoming connections.  The problem still occurs with the firewall switched off.
[moonlight-rtsp-error-50.pcapng.gz](https://github.com/moonlight-stream/moonlight-qt/files/6429951/moonlight-rtsp-error-50.pcapng.gz)

Thanks!

**Comment by cgutman** [comment]
Indeed the PCAP looks normal. Do you still get Error 50 with the old v3.1.1 version?

**Comment by mccoy-pauley-df** [comment]
Yes, the same Error 50 occurs now with 3.1.1 (despite the fact that it had been working previously).  Log is attached.
[Moonlight-1620252011.log](https://github.com/moonlight-stream/moonlight-qt/files/6430496/Moonlight-1620252011.log)

**Comment by mccoy-pauley-df** [comment]
Update -- no change in the problem, but did another Wireshark capture and noticed that there was more port 48010 traffic than the first one.  Not sure if it was different behavior or the second connection just got filtered out somehow.  Second file is attached, with lengthier conversation on source port 51155.

[moonlight-rtsp-error-50-take2.pcapng.gz](https://github.com/moonlight-stream/moonlight-qt/files/6442597/moonlight-rtsp-error-50-take2.pcapng.gz)

**Comment by mccoy-pauley-df** [comment]
OK, made some progress with this.  3.1.3 makes no difference to the problem, but I was able to produce different failure modes by disabling the Windows Defender firewall.  I've got a remote desktop entry defined using "mstsc.exe", so I tested with that, and with Windows firewall turned off.

When trying to start the Remote Desktop entry from Moonlight on Mac, there's a lengthy pause during "Starting RTSP handshake...", followed by Error 60.  Corresponding log entries:
﻿00:17:27 - SDL Info (0): Starting RTSP handshake...
00:17:37 - SDL Info (0): Connection timed out after 10 seconds (TCP port 48010)
00:17:37 - SDL Info (0): RTSP OPTIONS request failed: 60
00:17:37 - SDL Info (0): failed: 60

After the Remote Desktop "game" is started, trying to connect again results in a much quicker Error 50.  Wireshark shows that when I get error 60, the client is repeatedly sending a packet on 48000/udp and receiving an ICMP Port Unreachable packet in return, for about 10 seconds.  When I get error 50, the client sends one packet, gets an ICMP Port Unreachable, and immediately fails.  The pattern is consistent: Error 60 when starting the "game", Error 50 when resuming.  It appears that the nvstreamer process on the server is not creating a listener on 48000/udp when Moonlight on the Mac tries to connect to it.

However, when I try to connect to Remote Desktop from iPad using the iOS client, it immediately succeeds (even if I'm resuming the "game" started by Moonlight on Mac), and I can see using "netstat -aon" on the Windows server that the nvstreamer process is listening on port 48000/udp.

It seems like whenever Moonlight on Mac requests a connection, the server has some issue with the request and fails to open streaming on 48000/udp.  However, it has no problem with the same request coming from the Moonlight iOS client.

Here's a capture of port 48000/udp traffic from the Mac client.  Packets 1-30 are the initial attempt to connect resulting in Error 60.  Packets 31-32 are attempting to reconnect, giving the quick Error 50.
[Port-48000_No-Windows-FW_Start-and-Resume.pcapng.gz](https://github.com/moonlight-stream/moonlight-qt/files/6455230/Port-48000_No-Windows-FW_Start-and-Resume.pcapng.gz)

What else can I capture to help diagnose this?  Would love to be able to get Moonlight working again and stop having to rely on Steam Remote Play.

Thanks!

**Comment by cgutman** [comment]
I don't think the server opens the audio port (UDP 48000) until later in the RTSP handshake, so some ICMP Port Unreachable errors are expected there. I suspect the RTSP handshake just stalls before the point where UDP 48000 is opened, so this is a manifestation of the original issue.

The interesting change is that TCP 48010 is now failing with `ETIMEDOUT`, rather than that strange `ENETDOWN` error. It's still odd though, because you'd typically only see `ETIMEDOUT` if a firewall was dropping the SYN packets. Typically if the port wasn't open, you'd get a RST back from the server and fail immediately with `ECONNREFUSED`. And if the server process was hung, you'd end up with a socket that was "connected" then time out on the `recv()` but that's not what we see here either.

Please upload a packet capture with with `tcp.port == 47989 || tcp.port == 47984 || udp.port == 47998 || udp.port == 47999 || udp.port == 48000`

It also probably wouldn't hurt to upload the host-side logs too from `C:\ProgramData\NVIDIA Corporation\NvStream`

**Comment by mccoy-pauley-df** [comment]
Thanks -- was wondering where those Nvidia logs were. OK, ran the whole sequence again (start Moonlight, click on Remote Desktop to start it, wait for Error 60 to show up) while doing Wireshark captures on both client and server.  Those are attached, filtered down to the requested ports, along with the host-side logs from the NvStream directory, and the Moonlight log from the client.  They should all match the same session, having checked the timestamps (the matching streamer log should be NvStreamerCurrent.log).
[error60-failure-client-capture.pcapng.gz](https://github.com/moonlight-stream/moonlight-qt/files/6456203/error60-failure-client-capture.pcapng.gz)
[error60-failure-server-capture.pcapng.gz](https://github.com/moonlight-stream/moonlight-qt/files/6456204/error60-failure-server-capture.pcapng.gz)
[nvstreamerlogs.zip](https://github.com/moonlight-stream/moonlight-qt/files/6456205/nvstreamerlogs.zip)
[Moonlight-1620699877.log](https://github.com/moonlight-stream/moonlight-qt/files/6456206/Moonlight-1620699877.log)

I am a bit concerned that the only 48010 connection seems to be a SYN/RST at about the time that the failure is reported (hard to tell since Moonlight log has relative timestamps.)  Please let me know what else I can dig up if needed.

Thanks!

**Comment by mccoy-pauley-df** [comment]
Hokay.  Since it had become obvious that something was messing around at the kernel level, I finally went ahead and completely uninstalled Sophos Home antivirus, and that seems to have fixed the problem.  (I had tried disabling every piece of functionality previously, but that didn't seem to make a difference.)  So it looks like Sophos Home needs to go on the list of software that can cause problems with Moonlight.  I still have no idea why it suddenly started, since I've had Sophos on the Mac since the day I got it.  Will have to check out some alternatives to see if they work better.
Cameron, thank you for all your help and patience with this particularly aggravating issue! Moonlight is a great piece of software, and discovering it made me extremely happy!  I'm equally happy to have it working properly again.
